### ISSUE TITLE
Randomly getting a message to update the workload

### ISSUE THREAD

**Issue by raffaeler** [dotnet/aspire#1655]
When I start the Aspire host, I randomly get the following error (and the app stops):
```
info: Aspire.Dashboard.DashboardWebApplication[0]
      Now listening on: http://localhost:15267
info: Aspire.Dashboard.DashboardWebApplication[0]
      OTLP server running at: http://localhost:16276
Newer version of .NET Aspire workload is required to run the application. Run 'dotnet workload update' to get it.
```

If I re-run the app without even recompiling, it works.

Currently using Preview 2 bits with the updated workload as of today.

**Comment by davidfowl** [comment]
Can you show the output of `dotnet workload list`

**Comment by raffaeler** [comment]
@davidfowl Here it is:

```
>dotnet workload list

Installed Workload Id      Manifest Version                     Installation Source
-----------------------------------------------------------------------------------
wasm-tools                 8.0.1/8.0.100                        VS 17.9.34414.90
aspire                     8.0.0-preview.2.23619.3/8.0.100      VS 17.9.34414.90
```

**Comment by ElanHasson** [comment]
I'm using devcontainers and made [this devcontainer feature](https://github.com/ElanHasson/devcontainer-features/tree/main/src/dotnet-aspire-daily) for zero effort daily installs.  


I got a similar update workload error on some of the devcontainer builds, but during a `dotnet restore`. It's sporadic and does not happen 100% of the time.  Of course I also did it like 50 times while building the feature 😂 rebuilding did make it go away, but rebuilding daily to get latest bits.


Next time I get it I'll paste the output here if it's the same as it may be the same issue.

**Comment by davidfowl** [comment]
I'm not 100% sure why this is happening randomly, it's coming from [here](https://github.com/dotnet/aspire/blob/f954f614d45ea4085e94abb435ad6bcbb8280911/src/Aspire.Hosting/Dcp/DcpHostService.cs#L306).

Any ideas @bwateratmsft?

**Comment by bwateratmsft** [comment]
I wonder if this could be related to https://github.com/microsoft/usvc/issues/126, though it is so strange that it's not happening consistently.

**Comment by raffaeler** [comment]
> I wonder if this could be related to https://github.com/microsoft/usvc/issues/126, though it is so strange that it's not happening consistently.

FYI I have no visibility over the linked repository ...

**Comment by bwateratmsft** [comment]
@raffaeler In `%USERPROFILE%\.nuget\packages\microsoft.developercontrolplane.windows-amd64`, what versions do you see?

**Comment by raffaeler** [comment]
@bwateratmsft there is no such entry:
![image](https://github.com/dotnet/aspire/assets/5183152/f23aa286-1f22-4324-be59-f91c0ef40851)

**Comment by bwateratmsft** [comment]
Sorry, I gave the wrong path. That's where they used to be when packaged as a NuGet. The current location is here: `C:\Program Files\dotnet\packs\Aspire.Hosting.Orchestration.win-x64`

**Comment by raffaeler** [comment]
No worries, here is the name of the folder inside the new location:
```
8.0.0-preview.1.23557.2
```
I currently have `VS 17.9.0 Preview 4.0` .

If you believe this won't happen in Aspire Preview 3, I can live with that.

**Comment by bwateratmsft** [comment]
That's interesting, we would expect there to be an 8.0.0-preview.2... folder in there. What happens if you do `dotnet workload update --interactive` or `dotnet workload repair --interactive`?

**Comment by raffaeler** [comment]
Here is the full log of those two commands.
Please note that:
- during the `update` I was requested to elevate the dotnet installer
- the outcome of the `repair` is just what you can read below (no elevation or anything else).

```
**********************************************************************
** Visual Studio 2022 Developer Command Prompt v17.9.0-pre.4.0
** Copyright (c) 2022 Microsoft Corporation
**********************************************************************

C:\Program Files\Microsoft Visual Studio\2022\Preview>cd C:\Program Files\dotnet\packs\Aspire.Hosting.Orchestration.win-x64

C:\Program Files\dotnet\packs\Aspire.Hosting.Orchestration.win-x64>dir
 Volume in drive C has no label.
 Volume Serial Number is 6A09-FF05

 Directory of C:\Program Files\dotnet\packs\Aspire.Hosting.Orchestration.win-x64

05-Jan-24  12:04    <DIR>          .
05-Jan-24  12:04    <DIR>          ..
05-Jan-24  12:04    <DIR>          8.0.0-preview.1.23557.2
               0 File(s)              0 bytes
               3 Dir(s)  383,183,831,040 bytes free

C:\Program Files\dotnet\packs\Aspire.Hosting.Orchestration.win-x64>dotnet workload update --interactive

Welcome to .NET 8.0!
---------------------
SDK Version: 8.0.200-preview.23624.5

Telemetry
---------
The .NET tools collect usage data in order to help us improve your experience. It is collected by Microsoft and shared with the community. You can opt-out of telemetry by setting the DOTNET_CLI_TELEMETRY_OPTOUT environment variable to '1' or 'true' using your favorite shell.

Read more about .NET CLI Tools telemetry: https://aka.ms/dotnet-cli-telemetry

----------------
Installed an ASP.NET Core HTTPS development certificate.
To trust the certificate, run 'dotnet dev-certs https --trust'
Learn about HTTPS: https://aka.ms/dotnet-https

----------------
Write your first app: https://aka.ms/dotnet-hello-world
Find out what's new: https://aka.ms/dotnet-whats-new
Explore documentation: https://aka.ms/dotnet-docs
Report issues and find source on GitHub: https://github.com/dotnet/core
Use 'dotnet --help' to see available commands or visit: https://aka.ms/dotnet-cli
--------------------------------------------------------------------------------------

No workloads installed for this feature band. To update workloads installed with earlier SDK versions, include the --from-previous-sdk option.
Updated advertising manifest microsoft.net.workload.emscripten.net7.
Updated advertising manifest microsoft.net.workload.mono.toolchain.net7.
Updated advertising manifest microsoft.net.workload.emscripten.net6.
Updated advertising manifest microsoft.net.workload.mono.toolchain.current.
Updated advertising manifest microsoft.net.sdk.android.
Updated advertising manifest microsoft.net.workload.emscripten.current.
Updated advertising manifest microsoft.net.workload.mono.toolchain.net6.
Updated advertising manifest microsoft.net.sdk.ios.
Updated advertising manifest microsoft.net.sdk.maui.
Updated advertising manifest microsoft.net.sdk.macos.
Updated advertising manifest microsoft.net.sdk.tvos.
Failed to update the advertising manifest microsoft.net.sdk.aspire: Failed to validate package signing.

Verifying Microsoft.NET.Sdk.Aspire.Manifest-8.0.100.Msi.x64.8.0.0-preview.4.24079.1

error: NU3004: The package is not signed.

Package signature validation failed.

.
Updated advertising manifest microsoft.net.sdk.maccatalyst.
Downloading microsoft.net.sdk.android.manifest-8.0.100.msi.x64 (34.0.80)
Installing microsoft.net.sdk.android.manifest-8.0.100.msi.x64 .... Done
Downloading microsoft.net.sdk.ios.manifest-8.0.100.msi.x64 (17.2.8020)
Installing microsoft.net.sdk.ios.manifest-8.0.100.msi.x64 .... Done
Downloading microsoft.net.sdk.maccatalyst.manifest-8.0.100.msi.x64 (17.2.8020)
Installing microsoft.net.sdk.maccatalyst.manifest-8.0.100.msi.x64 .... Done
Downloading microsoft.net.sdk.macos.manifest-8.0.100.msi.x64 (14.2.8020)
Installing microsoft.net.sdk.macos.manifest-8.0.100.msi.x64 .... Done
Downloading microsoft.net.sdk.maui.manifest-8.0.100.msi.x64 (8.0.6)
Installing microsoft.net.sdk.maui.manifest-8.0.100.msi.x64 .... Done
Downloading microsoft.net.sdk.tvos.manifest-8.0.100.msi.x64 (17.2.8020)
Installing microsoft.net.sdk.tvos.manifest-8.0.100.msi.x64 .... Done
No workloads installed for this feature band. To update workloads installed with earlier SDK versions, include the --from-previous-sdk option.

Successfully updated workload(s): .


C:\Program Files\dotnet\packs\Aspire.Hosting.Orchestration.win-x64>dir
 Volume in drive C has no label.
 Volume Serial Number is 6A09-FF05

 Directory of C:\Program Files\dotnet\packs\Aspire.Hosting.Orchestration.win-x64

05-Jan-24  12:04    <DIR>          .
05-Jan-24  12:04    <DIR>          ..
05-Jan-24  12:04    <DIR>          8.0.0-preview.1.23557.2
               0 File(s)              0 bytes
               3 Dir(s)  383,170,129,920 bytes free

C:\Program Files\dotnet\packs\Aspire.Hosting.Orchestration.win-x64>dotnet workload repair --interactive

No workloads are installed, nothing to repair. Run `dotnet workload search` to find workloads to install.

C:\Program Files\dotnet\packs\Aspire.Hosting.Orchestration.win-x64>
```

Is the NuGet signing error the culprit?

**Comment by bwateratmsft** [comment]
@danegsta do you have any ideas? It's strange the workload doesn't seem to be fully installing and even stranger that the error is intermittent.

**Comment by bwateratmsft** [comment]
@raffaeler how often does this error occur?

**Comment by raffaeler** [comment]
It used to happen quite often. After you asked me to run the two command lines, now it always fails.
This is what I see in the output window:

```
Build started at 21:27...
1>------ Build started: Project: AspireStarter.ServiceDefaults, Configuration: Debug Any CPU ------
1>C:\Program Files\dotnet\sdk\8.0.200-preview.23624.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
1>Skipping analyzers to speed up the build. You can execute 'Build' or 'Rebuild' command to run analyzers.
1>AspireStarter.ServiceDefaults -> C:\dev.git\Experiments\Aspire\AspireStarter\AspireStarter.ServiceDefaults\bin\Debug\net8.0\AspireStarter.ServiceDefaults.dll
2>------ Build started: Project: AspireStarter.Web, Configuration: Debug Any CPU ------
3>------ Build started: Project: AspireStarter.ApiService, Configuration: Debug Any CPU ------
3>C:\Program Files\dotnet\sdk\8.0.200-preview.23624.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
2>Skipping analyzers to speed up the build. You can execute 'Build' or 'Rebuild' command to run analyzers.
2>AspireStarter.Web -> C:\dev.git\Experiments\Aspire\AspireStarter\AspireStarter.Web\bin\Debug\net8.0\AspireStarter.Web.dll
3>Skipping analyzers to speed up the build. You can execute 'Build' or 'Rebuild' command to run analyzers.
3>AspireStarter.ApiService -> C:\dev.git\Experiments\Aspire\AspireStarter\AspireStarter.ApiService\bin\Debug\net8.0\AspireStarter.ApiService.dll
4>------ Build started: Project: AspireStarter.AppHost, Configuration: Debug Any CPU ------
4>C:\Program Files\dotnet\sdk\8.0.200-preview.23624.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.ImportWorkloads.targets(38,5): error NETSDK1147: To build this project, the following workloads must be installed: aspire
4>C:\Program Files\dotnet\sdk\8.0.200-preview.23624.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.ImportWorkloads.targets(38,5): error NETSDK1147: To install these workloads, run the following command: dotnet workload restore
4>Done building project "AspireStarter.AppHost.csproj" -- FAILED.
========== Build: 3 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
========== Build completed at 21:28 and took 05.902 seconds ==========

```

**Comment by danegsta** [comment]
@raffaeler It looks like your machine has an advertising manifest for the main branch build of aspire (that would happen if you're running the `dotnet workload` commands from the Aspire repo).

If you want to install the latest main branch build of Aspire, you can follow the [instructions here](https://github.com/dotnet/aspire/blob/main/docs/using-latest-daily.md#set-up-your-machine-to-use-the-latest-aspire-builds).

From an elevated powershell terminal:
```powershell
dotnet workload update --skip-sign-check --source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json
dotnet workload install aspire --skip-sign-check --source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json
```

If you want the latest release build (preview 3 is currently being finalized) you can run the script from the Aspire repo [mentioned here](https://github.com/dotnet/aspire/blob/main/docs/using-latest-daily.md#optional-using-scripts-to-install-the-latest-net-aspire-build-from-release-branches).

From an elevated powershell terminal in the Aspire repo (with main or release/8.0-preview3 checked out):
```powershell
./eng/installLatestFromReleaseBranch.ps1
```

**Comment by raffaeler** [comment]
@danegsta This is a fresh-installed VM with VS2022 and VS2022 Preview.
I never cloned the Aspire repo on this machine. I cloned the Aspire-samples repo.
Finally, you can see that I repeated the workload commands from a folder which does not contain any repo and the result were the same.

The big question is: what else could have happened here? Apparently my machine does not fit the scenario you mentioned.

If you want me to run the scripts, please tell me if you prefer me to clone the Aspire repo and try the Preview 3 pre-release or the first two commands.

**Comment by danegsta** [comment]
@raffaeler dotnet maintains a local cache of available workload versions (pulled from nuget feeds) that can get updated by seemingly unrelated actions like restoring projects. I see a reference to version 8.0.0-preview.4.24079.1 of the Aspire workload (which is a recent main branch build). That means you've likely got a project somewhere that references the dotnet8 feed and restoring that unexpectedly caused your dotnet to pickup the main branch Aspire workload as the next available version.

If you don't mind running the latest development build, running those first two commands would be the fastest option for getting you unblocked. 

If you don't want to run a development build, a third option that doesn't require cloning the Aspire repo would be to use a rollback file to install a specific version of the workload:

Create a file named `aspire-rollback.txt` somewhere:
```json
{
  `"microsoft.net.sdk.aspire`": `"8.0.0-preview.2.23619.3/8.0.100`"
}
```

Then, from an elevated terminal in the same directory you created that file, run:
```powershell
dotnet workload update --source https://api.nuget.org/v3/index.json --from-rollback-file .\aspire-rollback.txt
dotnet workload install aspire --source https://api.nuget.org/v3/index.json --from-rollback-file .\aspire-rollback.txt
```

That'll force dotnet to install the current public preview 2 version of the workload from nuget.org.

**Comment by raffaeler** [comment]
@danegsta
>  I see a reference to version 8.0.0-preview.4.24079.1 of the Aspire workload (which is a recent main branch build). That means you've likely got a project somewhere that references the dotnet8 feed and restoring that unexpectedly caused your dotnet to pickup the main branch Aspire workload as the next available version.

Unless the `Aspire-samples` repo has some reference to it, I don't see how this could ever happen. I never cloned the main `Aspire` repo on this machine.


I just run the first two commands and this is the result:
```
C:\Program Files\Microsoft Visual Studio\2022\Preview>dotnet workload update --skip-sign-check --source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json

No workloads installed for this feature band. To update workloads installed with earlier SDK versions, include the --from-previous-sdk option.
Advertising manifest not updated. Manifest package for microsoft.net.workload.emscripten.current doesn't exist.
Advertising manifest not updated. Manifest package for microsoft.net.workload.mono.toolchain.current doesn't exist.
Skipping NuGet package signature verification.
Updated advertising manifest microsoft.net.sdk.maui.
Updated advertising manifest microsoft.net.sdk.android.
Updated advertising manifest microsoft.net.sdk.tvos.
Updated advertising manifest microsoft.net.sdk.macos.
Updated advertising manifest microsoft.net.workload.emscripten.net7.
Updated advertising manifest microsoft.net.workload.emscripten.net6.
Updated advertising manifest microsoft.net.sdk.aspire.
Updated advertising manifest microsoft.net.sdk.ios.
Updated advertising manifest microsoft.net.workload.mono.toolchain.net7.
Updated advertising manifest microsoft.net.workload.mono.toolchain.net6.
Updated advertising manifest microsoft.net.sdk.maccatalyst.
Downloading microsoft.net.sdk.aspire.manifest-8.0.100.msi.x64 (8.0.0-preview.4.24079.3)
Installing microsoft.net.sdk.aspire.manifest-8.0.100.msi.x64 .... Done
No workloads installed for this feature band. To update workloads installed with earlier SDK versions, include the --from-previous-sdk option.

Successfully updated workload(s): .


C:\Program Files\Microsoft Visual Studio\2022\Preview>dotnet workload install aspire --skip-sign-check --source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json

Advertising manifest not updated. Manifest package for microsoft.net.workload.mono.toolchain.current doesn't exist.
Advertising manifest not updated. Manifest package for microsoft.net.workload.emscripten.current doesn't exist.
Skipping NuGet package signature verification.
Downloading Aspire.Hosting.Sdk.Msi.x64 (8.0.0-preview.4.24079.3)
Installing Aspire.Hosting.Sdk.Msi.x64 .... Done
Downloading Aspire.ProjectTemplates.Msi.x64 (8.0.0-preview.4.24079.3)
Installing Aspire.ProjectTemplates.Msi.x64 .... Done
Downloading Aspire.Hosting.Orchestration.win-x64.Msi.x64 (8.0.0-preview.4.24079.3)
Installing Aspire.Hosting.Orchestration.win-x64.Msi.x64 ...... Done
Downloading Aspire.Hosting.Msi.x64 (8.0.0-preview.4.24079.3)
Installing Aspire.Hosting.Msi.x64 .... Done
Downloading Aspire.Dashboard.Sdk.win-x64.Msi.x64 (8.0.0-preview.4.24079.3)
Installing Aspire.Dashboard.Sdk.win-x64.Msi.x64 ..... Done

Successfully installed workload(s) aspire.

Unhandled exception.
C:\Program Files\Microsoft Visual Studio\2022\Preview>

```

The last `Unhandled exception` and the `8.0.0-preview.4.24079.3` version are not encouraging and in fact the rebuild all fails but with a different error:

```
Rebuild started at 22:46...
1>------ Rebuild All started: Project: StressApp, Configuration: Debug Any CPU ------
2>------ Rebuild All started: Project: AspireStarter.ServiceDefaults, Configuration: Debug Any CPU ------
Restored C:\dev.git\Experiments\Aspire\AspireStarter\AspireStarter.Web\AspireStarter.Web.csproj (in 206 ms).
Restored C:\dev.git\Experiments\Aspire\AspireStarter\AspireStarter.ApiService\AspireStarter.ApiService.csproj (in 206 ms).
Restored C:\dev.git\Experiments\Aspire\AspireStarter\AspireStarter.ServiceDefaults\AspireStarter.ServiceDefaults.csproj (in 4 ms).
Restored C:\dev.git\Experiments\Aspire\AspireStarter\StressApp\StressApp.csproj (in 17 ms).
Restored C:\dev.git\Experiments\Aspire\AspireStarter\AspireStarter.AppHost\AspireStarter.AppHost.csproj (in 221 ms).
2>C:\Program Files\dotnet\sdk\8.0.200-preview.23624.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
1>C:\Program Files\dotnet\sdk\8.0.200-preview.23624.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
2>AspireStarter.ServiceDefaults -> C:\dev.git\Experiments\Aspire\AspireStarter\AspireStarter.ServiceDefaults\bin\Debug\net8.0\AspireStarter.ServiceDefaults.dll
3>------ Rebuild All started: Project: AspireStarter.Web, Configuration: Debug Any CPU ------
4>------ Rebuild All started: Project: AspireStarter.ApiService, Configuration: Debug Any CPU ------
4>C:\Program Files\dotnet\sdk\8.0.200-preview.23624.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
1>StressApp -> C:\dev.git\Experiments\Aspire\AspireStarter\StressApp\bin\Debug\net8.0\StressApp.dll
3>AspireStarter.Web -> C:\dev.git\Experiments\Aspire\AspireStarter\AspireStarter.Web\bin\Debug\net8.0\AspireStarter.Web.dll
4>AspireStarter.ApiService -> C:\dev.git\Experiments\Aspire\AspireStarter\AspireStarter.ApiService\bin\Debug\net8.0\AspireStarter.ApiService.dll
5>------ Rebuild All started: Project: AspireStarter.AppHost, Configuration: Debug Any CPU ------
5>C:\dev.git\Experiments\Aspire\AspireStarter\AspireStarter.AppHost\Program.cs(11,46,11,70): error CS0234: The type or namespace name 'AspireStarter_ApiService' does not exist in the namespace 'Projects' (are you missing an assembly reference?)
5>C:\dev.git\Experiments\Aspire\AspireStarter\AspireStarter.AppHost\Program.cs(14,29,14,46): error CS0234: The type or namespace name 'AspireStarter_Web' does not exist in the namespace 'Projects' (are you missing an assembly reference?)
5>Done building project "AspireStarter.AppHost.csproj" -- FAILED.
========== Rebuild All: 4 succeeded, 1 failed, 0 skipped ==========
========== Rebuild completed at 22:46 and took 04.379 seconds ==========
```

This is even more unexpected ...

**Comment by danegsta** [comment]
Preview 3 introduced some breaking build changes (that are also in Preview 4 builds) that can cause issues if your project nugets aren't on Preview 3 or greater, particularly with regards to ProjectReference handling in the AppHost, which is likely what's causing those build failures. Since aspire-samples is still targeting 8.0.0-preview.2.23619.3 for the Aspire nuget packages it references, I'd recommend creating that workload rollback file I mentioned and using it to install preview 2 again.

**Comment by raffaeler** [comment]
@danegsta I created the file and run the first command with the elevated prompt:
```
C:\dev.git>dotnet workload update --source https://api.nuget.org/v3/index.json --from-rollback-file .\aspire-rollback.txt

Updated advertising manifest microsoft.net.workload.mono.toolchain.net7.
Updated advertising manifest microsoft.net.sdk.macos.
Updated advertising manifest microsoft.net.sdk.maccatalyst.
Updated advertising manifest microsoft.net.sdk.ios.
Updated advertising manifest microsoft.net.workload.emscripten.current.
Updated advertising manifest microsoft.net.workload.emscripten.net6.
Updated advertising manifest microsoft.net.workload.mono.toolchain.current.
Updated advertising manifest microsoft.net.workload.emscripten.net7.
Updated advertising manifest microsoft.net.workload.mono.toolchain.net6.
Updated advertising manifest microsoft.net.sdk.aspire.
Updated advertising manifest microsoft.net.sdk.android.
Updated advertising manifest microsoft.net.sdk.maui.
Updated advertising manifest microsoft.net.sdk.tvos.
Workload installation failed. Rolling back installed packs...
Installation rollback failed: Failed to verify authenticode signature for C:\ProgramData\dotnet\workloads\microsoft.net.sdk.aspire.manifest-8.0.100.msi.x64\8.0.0-preview.4.24079.3\cecef4140073a3a948b64581d7aa0eda-x64.msi. Error: 0x800b0100. No signature was present in the subject.
Workload update failed: Failed to verify authenticode signature for C:\ProgramData\dotnet\workloads\Aspire.Hosting.Sdk.Msi.x64\8.0.0-preview.4.24079.3\0342ec2c11552c143d61a2cdbfffd2b7-x64.msi. Error: 0x800b0100. No signature was present in the subject.

```

Given the above error, I didn't run the second command

**Comment by danegsta** [comment]
Can you add `--skip-sign-check` to the commands? It's likely failing because the installed workload isn't signed.

**Comment by raffaeler** [comment]
@danegsta it finally worked.
I believe the whole story around the workload versions should be easier to troubleshoot.

I have indeed the dotnet8 feed on the nuget sources as suggested by another repo. But I never found any doc suggesting that a Nuget source feed could cause all this trouble.

How can I avoid this for the future?

**Comment by danegsta** [comment]
In the short term, I would recommend removing the dotnet8 feed, since it would have only be necessary while the .NET 8 SDK was still is preview; you may also want to clear your dotnet workload advertising cache to be safe (`%userprofile%\.dotnet\sdk-advertising`) to make sure you don't still have references to the dev builds of Aspire there. That would keep the dotnet workload commands from trying to install preview 4 builds.

Long term, the .NET SDK team is working on several improvements to the workload experience that would include the ability to pin the workloads for a project via global.json.

**Comment by davidfowl** [comment]
Great work @danegsta and @bwateratmsft ! This is definitely painful @raffaeler, thanks for your patience! I'm going to close this issue as resolved for preview3. @danegsta if you can point to any issues with the future work it would be appreciated.

Again, thanks for the quick root causing here.

**Comment by raffaeler** [comment]
> you may also want to clear your dotnet workload advertising cache to be safe (`%userprofile%\.dotnet\sdk-advertising`)

Thank you @davidfowl and @danegsta, I removed the feed and deleted the folder. I hope I understood correctly.

**Comment by raffaeler** [comment]
@danegsta @davidfowl 
For your info, after upgrading to the latest preview of VS2022, it happened again.
I repeated this procedure specifying the Aspire preview 3 release, and I could fix that, but I didn't expect this to happen again in the first place.

Please also note that I did not re-added the nuget preview feed, so this happened with just the default nuget feed.

HTH


> Create a file named `aspire-rollback.txt` somewhere:
> 
> ```json
> {
>   "microsoft.net.sdk.aspire": "8.0.0-preview.2.23619.3/8.0.100"
> }
> ```
> 
> Then, from an elevated terminal in the same directory you created that file, run:
> 
> ```powershell
> dotnet workload update --source https://api.nuget.org/v3/index.json --from-rollback-file .\aspire-rollback.txt
> dotnet workload install aspire --source https://api.nuget.org/v3/index.json --from-rollback-file .\aspire-rollback.txt
> ```
> 
> That'll force dotnet to install the current public preview 2 version of the workload from nuget.org.
